cost/azure/idle_compute_instances/azure_idle_compute_instances.pt

```diff
@@ -0,0 +1,277 @@
+      result["averagecpu"] = final_average
+      results.push(result)
+    }
+  })
+EOS
+end
+
+
+###############################################################################
+# Policy
+###############################################################################
+
+policy "azure_resource_policy" do
+  validate $ds_combineddata do
+    summary_template "{{ rs_project_name }} (Account ID: {{ rs_project_id }}): Azure idle compute instances found"
+    detail_template <<-EOS
+# Azure idle compute instances
+
+| Region/Location | Name | CPU Average | Resource Group | Tags |
+| -------- | ---- | ----------- | -------------- | ---- |
+{{ range data -}}
+| {{ .location }} | {{.vmname}} | {{.averagecpu}} | {{.rg}} | {{range $i,$e := .tags}}{{$i}}:{{$e}},{{end}} |
+{{ end -}}
+
+EOS
+    check eq(size(data), 0)
+    escalate $email
+    escalate $delete_resources
+  end
+end
+
+###############################################################################
+# Escalations
+###############################################################################
+
+escalation "email" do
+   email $param_email
+end
+
+escalation "delete_resources" do
+  request_approval  do
+    label "Approve Resource Deletion"
+    description "Approve escalation to terminate the instance(s)"
+    parameter "approval_reason" do
+    type "string"
+      label "Reason for Approval"
+     description "Explain why you are approving the action"
+    end
+  end
+ run "delete_resources", data, $param_azure_tenant_id,$param_azure_sub
+end
+
+###############################################################################
+# Cloud Workflow
+###############################################################################
+
+define delete_resources($data, $param_azure_tenant_id,$param_azure_sub) return $all_responses do
+  $$debug=true
+  $$log = []
+  $all_responses = []
+  $syslog_subject = "Azure Terminate Instance: "
+  call sys_log(join([$syslog_subject, "Identified Instances"]),to_s($data))
+  call get_access_token($param_azure_tenant_id) retrieve $access_token
+  $all_responses = []
+  foreach $item in $data do
+   sub on_error: skip do
+      $response = http_request(
+        verb: "delete",
+        host: "management.azure.com",
+        href: join(["/subscriptions/", $param_azure_sub, "/resourceGroups/", $item["rg"], "/providers/Microsoft.Compute/virtualMachines/",$item["vmname"]]),
+        https: true,
+        query_strings: {
+          "api-version": "2018-06-01"
+        },
+        headers: {
+          "cache-control": "no-cache",
+          "content-type": "application/json",
+          "authorization": "Bearer " + $access_token
+        }
+      )
+      $all_responses << $response
+    end
+  end
+  call sys_log(join([$syslog_subject, "Responses"]),to_s($all_responses))
+end
+
+
+
+define get_access_token($param_azure_tenant_id) return $access_token do
+
+  $client_id = cred("AZURE_APPLICATION_ID")
+  $client_secret = cred("AZURE_APPLICATION_KEY")
```

Comment:
need to url_encode the secret.
https://github.com/flexera/policy_templates/blob/master/cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt#L511